server/src/main/java/org/apache/iotdb/db/metadata/mtree/traverser/counter/DatabaseCounter.java

```diff
@@ -22,19 +22,33 @@
 import org.apache.iotdb.commons.path.PartialPath;
 import org.apache.iotdb.db.metadata.mnode.IMNode;
 import org.apache.iotdb.db.metadata.mtree.store.IMTreeStore;
-import org.apache.iotdb.db.metadata.mtree.traverser.Traverser;
+import org.apache.iotdb.db.metadata.mtree.traverser.basic.DatabaseTraverser;
 
-// This class define the count as traversal result.
-public abstract class CounterTraverser extends Traverser {
+// This class implement database counter.
+public class DatabaseCounter extends DatabaseTraverser<Void> implements Counter {
 
-  protected long count;
+  private int count;
 
-  protected CounterTraverser(IMNode startNode, PartialPath path, IMTreeStore store)
+  public DatabaseCounter(
+      IMNode startNode, PartialPath path, IMTreeStore store, boolean isPrefixMatch)
       throws MetadataException {
-    super(startNode, path, store);
+    super(startNode, path, store, isPrefixMatch);
   }
 
-  public long getCount() {
+  @Override
+  protected Void generateResult(IMNode nextMatchedNode) {
+    count++;
+    return null;
+  }
+
+  @Override
+  public long count() throws MetadataException {
+    while (hasNext()) {
+      next();
+    }
+    if (getFailure() != null) {
+      throw new MetadataException(getFailure());
```

Comment:
Set the failure message to thrown MetadataException.

Reply:
done